File: Python/Clase08/capa_dato_persona/Persona.py
from logger_base import log  # Importamos el sistema de registro personalizado para el control de logs

# Definimos la clase Persona con atributos privados y métodos para acceder y modificar sus datos.
# Se implementa un constructor, el método especial __str__ para imprimir de forma legible,
# y propiedades con decoradores para manejar el acceso a los atributos.

class Persona:
    def __init__(self, id_persona=None, nombre=None, apellido=None, email=None):
        # Inicializamos los atributos del objeto. Todos los parámetros son opcionales.
        self._id_persona = id_persona
        self._nombre = nombre
        self._apellido = apellido
        self._email = email

    def __str__(self):
        # Devuelve una cadena con la información del objeto para facilitar su lectura.
        return f"""
        Id Persona: {self._id_persona},
        Nombre: {self._nombre},
        Apellido: {self._apellido},
        Email: {self._email}
    """

    # Accesores (getters) y modificadores (setters) definidos mediante @property
    # Esto permite un acceso seguro y controlado a los atributos de la clase.

    @property
    def id_persona(self):
        return self._id_persona

    @id_persona.setter
    def id_persona(self, id_persona):
        self._id_persona = id_persona

    @property
    def nombre(self):
        return self._nombre

    @nombre.setter
    def nombre(self, nombre):
        self._nombre = nombre

    @property
    def apellido(self):
        return self._apellido

    @apellido.setter
    def apellido(self, apellido):
        self._apellido = apellido

    @property
    def email(self):
        return self._email

    @email.setter
    def email(self, email):
        self._email = email

# Bloque de prueba: se ejecuta solo si el archivo es el principal
if __name__ == "__main__":
    # Creamos una instancia con todos los datos disponibles
    persona1 = Persona(1, "Juan", "Perez", "jperez@gmail.com")
    log.debug(persona1)  # Mostramos la información en el log

    # Otra instancia, esta vez sin especificar el ID
    persona2 = Persona(nombre="Jose", apellido="Lepez", email="ljose@gmail.com")
    log.debug(persona2)

    # Ejemplo de objeto creado solo con el identificador
    persona1 = Persona(id_persona=1)
    log.debug(persona1)

File: Java/Clase07/JavaBeans/src/domain/Persona.java
package domain;

import java.io.Serializable;

public class Persona implements Serializable {
    private String nombre;
    private String apellido;

    // Constructor sin argumentos (necesario para algunas operaciones como serialización o frameworks)
    public Persona() {
    }

    // Constructor que permite inicializar directamente los atributos nombre y apellido
    public Persona(String nombre, String apellido) {
        this.nombre = nombre;
        this.apellido = apellido;
    }

    // Getter para obtener el nombre de la persona
    public String getNombre() {
        return nombre;
    }

    // Setter para modificar el nombre
    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    // Getter para obtener el apellido
    public String getApellido() {
        return apellido;
    }

    // Setter para modificar el apellido
    public void setApellido(String apellido) {
        this.apellido = apellido;
    }

    // Método que devuelve una representación en texto del objeto Persona
    @Override
    public String toString() {
        return "Persona{" + "nombre=" + nombre + ", apellido=" + apellido + '}';
    }
}
